### PR TITLE
Support VPC Lattice as an event source

### PR DESCRIPTION
**Pull request by mbfreder**
*Issue #, if available:*   #639

*Description of changes:*
- Added support for VPC Lattice as an event source
- Include a sample app with a cloudformation template to test using SAM CLI

By submitting this pull request

- [ ] I confirm that my contribution is made under the terms of the Apache 2.0 license.
- [ ] I confirm that I've made a best effort attempt to update all relevant documentation.

### CODE REVIEW

**Review by deki — CHANGES_REQUESTED** [review]
Successfully deployed the sample but it's hard to follow and needs more documentation along with it. We should also enable auth and leverage https://github.com/okigan/awscurl to call it.

Looking at https://docs.aws.amazon.com/lambda/latest/dg/services-vpc-lattice.html#vpc-lattice-register V1 payload is the default, it makes sense for me to only support V2 but we should avoid that customers end up misconfigured.

**Comment by deki** [comment]
@dmahapatro you may want to take a look as well

File: aws-serverless-java-container-core/pom.xml [hunk 1]
```diff
@@ -63,6 +70,12 @@
             <version>6.2.3</version>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>org.projectlombok</groupId>
+            <artifactId>lombok</artifactId>
```
[the remark this answers is not in the document]
Comment (mbfreder) [reply aws-serverless-java-container-core/pom.xml]: Sure. Thanks

File: aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/ApacheCombinedServletLogFormatter.java [hunk 2]
```diff
@@ -125,7 +125,7 @@ public String format(ContainerRequestType servletRequest, ContainerResponseType
         logLineBuilder.append(" ");
         logLineBuilder.append(servletRequest.getRequestURI());
         logLineBuilder.append(" ");
-        logLineBuilder.append(servletRequest.getProtocol());
+        //logLineBuilder.append(servletRequest.getProtocol());
```
[the remark this answers is not in the document]
Comment (mbfreder) [reply aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/ApacheCombinedServletLogFormatter.java]: There's no way to get the protocol from the lattice payload (confirmed with the VPC Lattice team), so I returned an UnsupportedOperationException which throws a nullpointerException on that line. Is it the right thing to return? or should we default to "HTTP/1.1" ?

File: samples/springboot3/vpc-lattice-pet-store/pom.xml [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+    <artifactId>serverless-springboot3-example</artifactId>
+    <version>2.0-SNAPSHOT</version>
+    <name>Spring Boot example for the aws-serverless-java-container library</name>
+    <description>Simple pet store written with the Spring framework and Spring Boot</description>
+    <url>https://aws.amazon.com/lambda/</url>
+
+    <parent>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-parent</artifactId>
+        <version>3.2.3</version>
```
[the remark this answers is not in the document]
Comment (mbfreder) [reply samples/springboot3/vpc-lattice-pet-store/pom.xml]: Thanks

File: samples/springboot3/vpc-lattice-pet-store/build.gradle [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+repositories {
+    mavenLocal()
+    mavenCentral()
+    maven {url "https://repo.spring.io/milestone"}
+    maven {url "https://repo.spring.io/snapshot"}
+}
+
+dependencies {
+    implementation (
+            implementation('org.springframework.boot:spring-boot-starter-web:3.2.3') {
```
[the remark this answers is not in the document]
Comment (mbfreder) [reply samples/springboot3/vpc-lattice-pet-store/build.gradle]: Thanks

File: samples/springboot3/vpc-lattice-pet-store/app/index.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+import os
+import json
+import http.client
+
+def handler(event, context):
```
[the remark this answers is not in the document]
Comment (mbfreder) [reply samples/springboot3/vpc-lattice-pet-store/app/index.py:5]: This is the code for the Outbound lambda function, which is used to call the SpringBoot application. I updated the README with a diagram to illustrate the flow.

File: samples/springboot3/vpc-lattice-pet-store/template.yaml [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+                  - ec2:DescribeNetworkInterfaces
+                  - ec2:DeleteNetworkInterface
+                Resource: '*'
+
+  OutboundLambdaFunction:
+    Type: AWS::Serverless::Function
+    Properties:
+      Handler: index.handler
+      Role: !GetAtt OutboundLambdaFunctionRole.Arn
+      Runtime: python3.9
```
[the remark this answers is not in the document]
Comment (mbfreder) [reply samples/springboot3/vpc-lattice-pet-store/template.yaml:167]: This is the client lambda function. I used python because it was much easier to write.

File: aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsVpcLatticeV2HttpServletRequest.java [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+import java.util.*;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * Implementation of the <code>HttpServletRequest</code> interface that supports <code>VPCLatticeV2RequestEvent</code> object.
+ * This object is initialized with an <code>VPCLatticeV2RequestEvent</code> event and a <code>SecurityContext</code> generated
+ * by an implementation of the <code>SecurityContextWriter</code>.
+ */
+public class AwsVpcLatticeV2HttpServletRequest extends AwsHttpServletRequest {
```
[the remark this answers is not in the document]
Comment (mbfreder) [reply aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsVpcLatticeV2HttpServletRequest.java:32]: I already moved most of the duplicated code in `AwsHttpServletRequest`.

File: aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/ApacheCombinedServletLogFormatter.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (deki) [reply aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/ApacheCombinedServletLogFormatter.java]: I'd default to HTTP as this is the only protocol supported anyway.
